# ----------------------------4.5: Data Consistency Checks------------------

# -Excercice

###### importing panda , numpy , os , df : prods, ords

In [3]:
import pandas as pd
import numpy as np
import os


In [4]:
path = r'C:\Users\bouas\instacart bascket analysis'
df_prods = pd.read_csv(os.path.join(path, 'data', 'original_data', 'products.csv'), index_col = False)


In [5]:
df_ords = pd.read_csv(os.path.join(path, 'data', 'prepared_data', 'orders_wrangled.csv'), index_col = False)


In [6]:
# mixed value check
df_test = pd.DataFrame()
df_test['mix'] = ['a', 'b', 1, True]

In [7]:
df_test.head()

,mix
0,a
1,b
2,1
3,True


In [10]:
print(df_test)


    mix
0     a
1     b
2     1
3  True


In [12]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [13]:
df_nan = df_prods[df_prods['product_name'].isnull()==True]

In [14]:
# displaying missing value
df_nan

,product_id,product_name,aisle_id,department_id,prices
33,34,NaN,121,14,12.2
68,69,NaN,26,7,11.8
115,116,NaN,93,3,10.8
261,262,NaN,110,13,12.1
525,525,NaN,109,11,1.2
1511,1511,NaN,84,16,14.3
1780,1780,NaN,126,11,12.3
2240,2240,NaN,52,1,14.2
2586,2586,NaN,104,13,12.4
3159,3159,NaN,126,11,13.1


In [19]:
df_nan.tail(10)

,product_id,product_name,aisle_id,department_id,prices
1780,1780,NaN,126,11,12.3
2240,2240,NaN,52,1,14.2
2586,2586,NaN,104,13,12.4
3159,3159,NaN,126,11,13.1
3230,3230,NaN,120,16,14.4
3736,3736,NaN,41,8,14.8
4283,4283,NaN,77,7,14.4
4790,4790,NaN,91,16,14.5
38187,38183,NaN,39,12,20.9
40444,40440,NaN,120,16,14.8


In [15]:
df_nan.describe().round(1)

,product_id,aisle_id,department_id,prices
count,16.0,16.0,16.0,16.0
mean,6684.0,89.9,10.9,13.0
std,12836.7,33.7,4.6,3.9
min,34.0,26.0,1.0,1.2
25%,459.2,70.8,7.8,12.2
50%,2413.0,98.5,11.5,13.6
75%,3872.8,120.0,14.5,14.4
max,40440.0,126.0,16.0,20.9


#### Investigation : The maximum of product_id (40440) is abnormally high, well beyond the other values (50%;25% AND 75%) after depp investigation we noticed that the last 2 rows are outliers when we compare it to rest of the other value (38183, 40440) 
#### the same for the price (20,9) it can be outlier because when we compare it to the other value (75,50,25%) it s abnormal
#### >>> so we can delete these 2 rows

In [21]:
df_prods.drop([38187, 40444], inplace=True)



In [23]:
df_nan.describe().round(1)

,product_id,aisle_id,department_id,prices
count,14.0,14.0,14.0,14.0
mean,2022.9,91.4,10.5,12.3
std,1661.5,32.3,4.8,3.4
min,34.0,26.0,1.0,1.2
25%,327.8,78.8,7.2,12.1
50%,2010.0,98.5,11.0,12.8
75%,3212.2,117.5,13.8,14.4
max,4790.0,126.0,16.0,14.8


#### Here is the values seems normal 

In [24]:
df_nan.tail(10)

,product_id,product_name,aisle_id,department_id,prices
525,525,NaN,109,11,1.2
1511,1511,NaN,84,16,14.3
1780,1780,NaN,126,11,12.3
2240,2240,NaN,52,1,14.2
2586,2586,NaN,104,13,12.4
3159,3159,NaN,126,11,13.1
3230,3230,NaN,120,16,14.4
3736,3736,NaN,41,8,14.8
4283,4283,NaN,77,7,14.4
4790,4790,NaN,91,16,14.5


In [34]:

df_prods.dropna(subset=['product_name'], inplace=True)

In [35]:
df_prods.shape

(49677, 5)

In [36]:
# displaying duplicate
df_prods_clean = df_prods.copy()
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [37]:
df_dups

,product_id,product_name,aisle_id,department_id,prices
462,462,Fiber 4g Gummy Dietary Supplement,70,11,4.8
18459,18458,Ranger IPA,27,5,9.2
26810,26808,Black House Coffee Roasty Stout Beer,27,5,13.4
35309,35306,Gluten Free Organic Peanut Butter & Chocolate ...,121,14,6.8
35495,35491,Adore Forever Body Wash,127,11,9.9


In [73]:
df_prods_clean.shape

(49677, 5)

In [80]:
# dropping duplicate
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [75]:
df_prods_clean_no_dups

,product_id,product_name,aisle_id,department_id,prices
0,1,Chocolate Sandwich Cookies,61,19,5.8
1,2,All-Seasons Salt,104,13,9.3
2,3,Robust Golden Unsweetened Oolong Tea,94,7,4.5
3,4,Smart Ones Classic Favorites Mini Rigatoni Wit...,38,1,10.5
4,5,Green Chile Anytime Sauce,5,13,4.3
...,...,...,...,...,...
49688,49684,"Vodka, Triple Distilled, Twist of Vanilla",124,5,5.3
49689,49685,En Croute Roast Hazelnut Cranberry,42,1,3.1
49690,49686,Artisan Baguette,112,3,7.8
49691,49687,Smartblend Healthy Metabolism Dry Cat Food,41,8,4.7


# -----------------------------------------[TASK]----------------------------------------

# 2/Run the df.describe() function on your df_ords dataframe

In [44]:

df_ords.describe().round(1)

,Unnamed: 0,order_id,user_id,order_number,order_dow,order_hour_of_day,days_since_prior_order
count,3421083.0,3421083.0,3421083.0,3421083.0,3421083.0,3421083.0,3214874.0
mean,1710541.0,1710542.0,102978.2,17.2,2.8,13.5,11.1
std,987581.7,987581.7,59533.7,17.7,2.0,4.2,9.2
min,0.0,1.0,1.0,1.0,0.0,0.0,0.0
25%,855270.5,855271.5,51394.0,5.0,1.0,10.0,4.0
50%,1710541.0,1710542.0,102689.0,11.0,3.0,13.0,7.0
75%,2565811.5,2565812.5,154385.0,23.0,5.0,16.0,15.0
max,3421082.0,3421083.0,206209.0,100.0,6.0,23.0,30.0


#### Investigation : 
##### Overall, the data looks consistent, but the missing values in `days_since_prior_order(` has fewer entries (≈3.22M vs. 3.42M))
##### the high maximum in `order_number`look strange
##### and the max since last order is also strange : 30 days 
#### >>>>all this info should be more investigated 

# 3/Check for mixed-type data in your df_ords dataframe.

In [50]:
for col in df_ords.columns.tolist():
    weird = (df_ords[col].map(type) != df_ords[col].iloc[0].__class__).any()
    if weird:
        print(col)

Unnamed: 0
order_id
user_id
order_number
order_dow
order_hour_of_day
days_since_prior_order


# 4/The column in question should contain observations of a single data type.

#### There is no observation of a single type

# 5/Run a check for missing values in your df_ords dataframe.


In [45]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

#### As we can see here and as we investigated in the previous check  the days since prior order has 206209 missing value

# 6/Address the missing values using an appropriate method

#### Given that the maximum value is strange, I think imputation by mean won't be the right solution; removing them isn't the right solution either. imputing by median will affect to the negative side the results, so it s better to do anything

##### Let's start by displaying the missing values.

In [30]:
df_NAN = df_ords[df_ords['days_since_prior_order'].isnull()==True]

In [31]:
df_NAN

,Unnamed: 0,order_id,user_id,eval_set,order_number,order_dow,order_hour_of_day,days_since_prior_order
0,0,2539329,1,prior,1,2,8,NaN
11,11,2168274,2,prior,1,2,11,NaN
26,26,1374495,3,prior,1,1,14,NaN
39,39,3343014,4,prior,1,6,11,NaN
45,45,2717275,5,prior,1,3,12,NaN
...,...,...,...,...,...,...,...,...
3420930,3420930,969311,206205,prior,1,4,12,NaN
3420934,3420934,3189322,206206,prior,1,3,18,NaN
3421002,3421002,2166133,206207,prior,1,6,19,NaN
3421019,3421019,2227043,206208,prior,1,1,15,NaN


# 7/Run a check for duplicate values in your df_ords data.

#####  Creating subset of df_ords witch have only the duplicates

In [59]:
df_ords_dups = df_ords[df_ords.duplicated()]

In [60]:
df_ords_dups


,Unnamed: 0,order_id,user_id,eval_set,order_number,order_dow,order_hour_of_day,days_since_prior_order


### There are no duplicates
###### if there dup, i will use drop_duplicates()

# 9/Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [32]:
df_ords.to_csv(os.path.join(path, 'data','prepared_data', 'orders_consistency.csv'))

In [33]:
df_prods.to_csv(os.path.join(path, 'data','prepared_data', 'prods_consistency.csv'))